# Phân Tích Hồ Sơ Ứng Dụng Có Lợi Nhuận trên App Store và Google Play

# _Profitable App Profiles for the App Store and Google Play Markets_

_This project is to generate foundation for readers understanding the mean of dataset_

## OPEN AND EXPLORE DATA

In [24]:
#Applestore
opened_file_1 = open('AppleStore.csv', encoding='utf-8')

from csv import reader
read_file_1 = reader(opened_file_1)
ios_apps = list(read_file_1)
ios_header = ios_apps[0]
ios_apps = ios_apps[1:]

#Googlepplay store
opened_file_2 = open('googleplaystore.csv', encoding='utf-8')

from csv import reader
read_file_2 = reader(opened_file_2)
android_apps = list(read_file_2)
android_header = android_apps[0]
android_apps = android_apps[1:]

In [32]:
#Using Explore function:

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line after each row

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
print(android_header)
print('\n')
explore_data(android_apps, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [34]:
#Print the incorrect row
print(android_apps[0])
print('\n')
print(android_header)
print('\n')
print(android_apps[10472])

#delete wrong data 10472
print(len(android_apps))
del android_apps[10472]
print(len(android_apps))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']
10841
10840


## REMOVING DUPLICATE APPS

In [36]:
#Part 1: we have some duplicate apps in this dataset
for app in android_apps:
    name = app[0]
    if name == 'Facebook':
        print(app)

# Count the nummber of duplicate apps:
duplicate_apps = []
unique_apps = []
for app in android_apps:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
print('Number of duplicate apps: ', len(duplicate_apps))
print('Example of duplicate apps:')
print(duplicate_apps[:5])
#### So we cannot remove randomly row in dataset, I'll use a common rule: only keep the newest data

['Facebook', 'SOCIAL', '4.1', '78158306', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
['Facebook', 'SOCIAL', '4.1', '78128208', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'August 3, 2018', 'Varies with device', 'Varies with device']
Number of duplicate apps:  1181
Example of duplicate apps:
['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


In [38]:
#Part 2: removing duplicate apps
#Step 1:
reviews_max = {}
for app in android_apps:
    name = app[0]
    n_reviews = float(app[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

#Step 2:
android_clean = []
already_added = []
for app in android_apps:
    name = app[0]
    n_reviews = float(app[3])
    if (n_reviews == reviews_max[name]) and (name not in already_added):
            android_clean.append(app) 
            already_added.append(name)
print(len(android_clean))
print(len(reviews_max))

9659
9659


## REMOVE NON ENGLISH APPS

In [42]:
#Generate new function:
def check_ea(string):
    count = 0
    for character in string:
        
        if ord(character) > 127:
            count += 1
    if count > 3:
        return False
    else:
        return True
        
#Checking:
print(check_ea('Instagram'))
print(check_ea('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(check_ea('Docs To Go™ Free Office Suite'))
print(check_ea('Instachat 😜'))

#Checking both dataset:
English_apps = []
Non_English_apps = []
for app in android_clean:
    name = app [0]
    if check_ea(name) == True:
        English_apps.append(app)
    else:
        Non_English_apps.append(app)
print(Non_English_apps[:2])
print(len(English_apps))

ios_english_apps = []
non_ios_english_apps = []
for app in ios_apps:
    name = app[1]
    if check_ea(name) == True:
        ios_english_apps.append(app)
    else:
        non_ios_english_apps.append(app)
print(len(ios_english_apps))

True
False
True
True
[['Flame - درب عقلك يوميا', 'EDUCATION', '4.6', '56065', '37M', '1,000,000+', 'Free', '0', 'Everyone', 'Education', 'July 26, 2018', '3.3', '4.1 and up'], ['သိင်္ Astrology - Min Thein Kha BayDin', 'LIFESTYLE', '4.7', '2225', '15M', '100,000+', 'Free', '0', 'Everyone', 'Lifestyle', 'July 26, 2018', '4.2.1', '4.0.3 and up']]
9614
6183


## ISOLATING FREE APPS

In [51]:
android_final = []
for app in English_apps:
    if app[7] == '0':
        android_final.append(app)
print(len(android_final))

ios_final = []
for app in ios_english_apps:
    if app[4] == '0.0':
        ios_final.append(app)
print(len(ios_final))

8864
3222


## DETERMINING COMMON APPS
As we mentioned in the introduction, our goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue.

To minimize risks and overhead, our validation strategy for an app idea has three steps:

- Build a minimal Android version of the app, and add it to Google Play.
- If the app has a good response from users, we develop it further.
- If the app is profitable after six months, we build an iOS version of the app and add it to the App Store.
Because our end goal is to add the app on both Google Play and the App Store, we need to find app profiles that are successful in both markets. For instance, a profile that works well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by determining the most common genres for each market. For this, we'll need to build frequency tables for 'Category' and 'Genres' on android_final ,and 'prime_genre' on ios_final

In [54]:
print(android_header)
print(android_final[0])
print('\n')
print(ios_header)
print(ios_final[0])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


In [56]:
def freq_table(dataset, index):
    freqtable = {}
    total = 0
    freqtable_percent = {}
    for app in dataset:
        total += 1
        name = app[index]
        if name in freqtable:
            freqtable[name] += 1
        else:
            freqtable[name] = 1
    for name in freqtable:
        freqtable_percent[name] = round((freqtable[name]/total) * 100, 2) 
    return freqtable_percent
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

### Common apps on IOS

In [61]:
display_table(ios_final, -5)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


- Most common genre: Games (58.16%), followed by Entertainment (7.88%).
- Pattern: Entertainment-focused apps dominate, while practical apps (education, shopping, utilities) - are fewer.
- General impression: The App Store market is largely geared towards entertainment rather than practical use.
- Recommendation: A gaming or entertainment app could succeed but faces high competition. Exploring niche categories like health & fitness or education might offer growth with less competition.
- App frequency vs. user base: A large number of apps in a genre doesn’t guarantee a large user base, but rather reflects a competitive market. Niche categories may have untapped potential.

### Common apps on Android

In [70]:
display_table(android_final, 1) # Category

FAMILY : 18.91
GAME : 9.72
TOOLS : 8.46
BUSINESS : 4.59
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.53
SPORTS : 3.4
PERSONALIZATION : 3.32
COMMUNICATION : 3.24
HEALTH_AND_FITNESS : 3.08
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.79
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.16
ENTERTAINMENT : 0.96
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


Following by Category
- Most common genres on Google Play: Family (18.91%), Game (9.72%), Tools (8.46%).
- Patterns: Google Play focuses more on practical apps (family, tools, business) compared to the App Store, which is dominated by entertainment, especially games.
- Comparison: Google Play has more diversity and utility-based apps, while the App Store is entertainment-centric.
- App profile recommendation: For Google Play, a Family or Tools app would align with the largest categories. For the App Store, gaming or entertainment apps fit the market but face high competition.
- Frequency vs. users: High app frequency in a genre suggests competition, not necessarily a large user base. Niche categories may offer growth with less competition.

In [76]:
display_table(android_final, -4) #Genres

Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.59
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.53
Sports : 3.46
Personalization : 3.32
Communication : 3.24
Action : 3.1
Health & Fitness : 3.08
Photography : 2.94
News & Magazines : 2.8
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.85
Video Players & Editors : 1.77
Casual : 1.76
Maps & Navigation : 1.4
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Role Playing : 0.94
Libraries & Demo : 0.94
Auto & Vehicles : 0.93
Strategy : 0.91
House & Home : 0.82
Weather : 0.8
Events : 0.71
Adventure : 0.68
Comics : 0.61
Beauty : 0.6
Art & Design : 0.6
Parenting : 0.5
Card : 0.45
Casino : 0.43
Trivia : 0.42
Educational;Education : 0.39
Board : 0.38
Educational : 0.37
Education;Education : 0.34
Word : 0.26
Casual;Pretend Play : 0.24
Music : 0.2
Racing;Action & Adventure : 0.17
Puzzle;Brain Games : 0.17
Entertainment;Music & Video : 0.17
Casual;

When we observe types of apps on GG play sorted by Genres, a bigger picture is provided with various details
=> Tools, entertainment, education, business are still the most popular apps, but this data can be useful when we combine them with data sorted by Catergory.
=> We can use genres to determine more detail for category.

In [81]:
freq_genre = freq_table(ios_final, -5)
for genre in freq_genre:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:
            total += float(app[5])
            len_genre += 1
    average = total/len_genre
    print(genre, ':', average)

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


## A frequency table for the Category column of the Google Play

In [85]:
freqtable_category = freq_table(android_final, 1)
for category in freqtable_category:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            install = app[5]
            remove_1 = install.replace('+', '')
            remove_2 = remove_1.replace(',', '')
            total += float(remove_2)
            len_category += 1
    average_gg = total/len_category
    print(category, ':', average_gg)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

Communication have around 38m users, the highest number of installs. Nervertheless, this field are obsersing highly competition from big firm: FB, GG, X,...

In [91]:
for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure : 100,000,000+
Viber Mess

We can see, each apps in this list serving their own duty and basing on local market, we can know that whether it is protential thriving or not.
For example, WhatsApp is very popular with 1m user, but in VN this app is quite not common, peopl